### PR TITLE
[libc++][format] P3142R0: Printing Blank Lines with `println`

### LINKED ISSUE #105415: P3142R0: Printing Blank Lines with `println`
**Link:** https://wg21.link/P3142R0

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements https://wg21.link/P3142R0

Applied retroactively as DR, same as stdlibc++ and MS STL:
https://github.com/orgs/microsoft/projects/1143?pane=issue&itemId=57457187

### CODE REVIEW

File: libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp [hunk 1]
```diff
@@ -86,10 +87,11 @@ test(std::stringstream& stream, std::string expected, test_format_string<char, A
...
   {
-    expected += '\n'; // Tested last since it changes the expected value.
+    expected += "\n\n"; // Tested last since it changes the expected value.
     stream.str("");
     ;
     std::println(stream, fmt, std::forward<Args>(args)...);
+    std::println(stream);
     std::string out = stream.str();
     TEST_REQUIRE(out == expected,
                  TEST_WRITE_CONCATENATED(
```
Comment (mordante) [inline libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp:97]: Please split in two tests, now if
` std::println(stream, fmt, std::forward<Args>(args)...); -> "\n\n"`
` std::println(stream)-> ""`
the test passes.

The same for the other tests, except the `println.sh.cpp`. That last test is a bit trickier so let's keep that as is.

**Review by mordante — CHANGES_REQUESTED** [review]
This looks quite good. A few minor comments. Can you rebase so you can update the status page.

File: libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp (hunk 1 above)
Comment (H-G-Hristov) [reply libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp]: Or we can have println_blank_line.sh.cpp?

Comment (mordante) [reply libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp]: I'm fine with an additional test too.

File: libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp [hunk 2]
```diff
@@ -95,6 +96,16 @@ test(std::stringstream& stream, std::string expected, test_format_string<char, A
...
                      "\nFormat string   ", fmt.get(), "\nExpected output ", expected, "\nActual output   ", out, '\n'));
   }
+  // *** println blank line***
+  {
+    expected = "\n";
+    stream.str("");
+    ;
```
Comment (mordante) [inline libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp:103]:
```suggestion
  
```
Can you also fix this on line 71 and 81?

Comment (Zingam) [reply libcxx/test/std/input.output/iostream.format/output.streams/ostream.formatted/ostream.formatted.print/locale-specific_form.pass.cpp:103]: Sure!

**Review by mordante — APPROVED** [review]
Thanks! LGTM!

**Comment by Zingam** [comment]
@mordante Thank you!

**Comment by vvereschaka** [comment]
@H-G-Hristov ,

the `llvm-libc++-static.cfg.in::println.blank_line.sh.cpp` test gets failed on the windows to linux cross builders with the following messages
```
diff -u C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\input.output\iostream.format\print.fun\Output\println.blank_line.sh.cpp.dir\t.tmp.actual C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\input.output\iostream.format\print.fun\Output\println.blank_line.sh.cpp.dir\t.tmp.expected
# executed command: diff -u 'C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\input.output\iostream.format\print.fun\Output\println.blank_line.sh.cpp.dir\t.tmp.actual' 'C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\input.output\iostream.format\print.fun\Output\println.blank_line.sh.cpp.dir\t.tmp.expected'
# .---command stdout------------
# | --- C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\input.output\iostream.format\print.fun\Output\println.blank_line.sh.cpp.dir\t.tmp.actual
# | +++ C:\buildbot\as-builder-1\x-armv7l\build\runtimes\runtimes-armv7-unknown-linux-gnueabihf-bins\test\std\input.output\iostream.format\print.fun\Output\println.blank_line.sh.cpp.dir\t.tmp.expected
# | @@ -1 +1 @@
# | -
# | +n
# `-----------------------------
# error: command failed with exit status: 1
```
* https://lab.llvm.org/buildbot/#/builders/60/builds/16712
* https://lab.llvm.org/buildbot/#/builders/119/builds/17020

would you take care of it?

**Comment by mordante** [comment]
I had a look and a similar test `std/input.output/iostream.format/print.fun/println.sh.cpp` passes on this platform. Maybe the issue is that the output only contains `\n`. Maybe changing the test to
```
print("hello world");
println();
```
might solve the issue. @Zingam can you provide a followup patch to make this change?
@vvereschaka would you be able to test that followup patch? Do you feel this is so urgent it warrants a temporary revert?

If the suggested solution works we should add comment why we output we do that.

**Pull request merged**